R/transactions.R

```diff
@@ -43,4 +46,71 @@ setGeneric("dbCommit",
 setGeneric("dbRollback",
   def = function(conn, ...) standardGeneric("dbRollback"),
   valueClass = "logical"
-)
\ No newline at end of file
+)
+
+#' Self-contained SQL transactions.
+#'
+#' Given that \code{\link{transactions}} are implemented, then this function
+#' allows you pass in code that is treated as a transaction. The advantage is
+#' that you don't have to remember to do \code{dbBegin} and \code{dbCommit} or
+#' \code{dbRollback} -- that is all taken care of.
+#'
+#' @section Side Effects:
+#' The transaction in \code{code} on the connection \code{conn} is committed
+#' or rolled back. The \code{code} chunk may also also modify the local R
+#' environment.
+#'
+#' @param conn A \code{\linkS4class{DBIConnection}} object, as produced by
+#'   \code{\link{dbConnect}}.
+#' @param code An arbitrary block of R code
+#'
+#' @return The result of the evaluation of \code{code}
+#' @aliases dbWithTransaction,DBIConnection-method
+#' @export
+#' @examples
+#' con <- dbConnect(RSQLite::SQLite(), ":memory:")
+#' dbWriteTable(con, "cars", head(cars, 3))
+#' dbReadTable(con, "cars")   # there's 3 rows!
+#' ## successful transaction
+#' dbWithTransaction(con, {
+#'   dbExecute(con, "INSERT INTO cars (speed, dist) VALUES (1, 1);")
+#'   dbExecute(con, "INSERT INTO cars (speed, dist) VALUES (2, 2);")
+#'   dbExecute(con, "INSERT INTO cars (speed, dist) VALUES (3, 3);")
+#' })
+#' dbReadTable(con, "cars")   # there's now 6 rows!
+#'
+#' \dontrun{
+#' ## unsuccessful transaction -- note the missing comma
+#' dbWithTransaction(con,{
+#'   dbExecute(con, "INSERT INTO cars (speed, dist) VALUES (1, 1);")
+#'   dbExecute(con, "INSERT INTO cars (speed dist) VALUES (2, 2);")
+#'   dbExecute(con, "INSERT INTO cars (speed, dist) VALUES (3, 3);")
+#' })
+#' dbReadTable(con, "cars")   # nothing was changed
+#' }
+#' dbDisconnect(con)
+setGeneric("dbWithTransaction",
+  def = function(conn, code) standardGeneric("dbWithTransaction")
+)
+
+#' @export
+setMethod("dbWithTransaction", "DBIConnection", function(conn, code) {
+  ## check if each operation is successful
+  if (!dbBegin(conn)) {
```

Comment:
Right, of course, if we stop inling `call`, there's no need to check for the `error` class... So, if we're not doing that, what exactly do you want `isFALSE()` to check for? In other words why isn't one of these sufficient:

```r
call <- dbBegin(conn)
if (!call) stop("`dbBegin` was not successful")
```
or
```r
call <- dbBegin(conn)
if (identical(call, FALSE)) stop("`dbBegin` was not successful")
```
There can't be any backend out there that, in case of error during `dbBegin`, doesn't either throw an error (in which case, as you mentioned, the code below is never reached) or at least returns `FALSE`, right? So I guess I'm not sure what else we're supposed to check for...